### PR TITLE
fix(Notion Node): Fix issue preventing some database page urls from working

### PR DESCRIPTION
**Pull request by Joffcom**
## Summary
Fixes an issue with the RLC regex that was incorrectly flagging `https://www.notion.so/fake-instance/x-ID` as invalid. Also added test for `extractPageId` that uses the regex pattern first.

## Related Linear tickets, Github issues, and Community forum posts
https://community.n8n.io/t/notion-update-by-url-not-working-for-guest-added-pages/50095

### CODE REVIEW

File: packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts [hunk 1]
```diff
@@ -601,15 +601,15 @@ export const databasePageFields: INodeProperties[] = [
 						type: 'regex',
 						properties: {
 							regex:
-								'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{2,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12}).*',
+								'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{1,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12}).*',
```
Comment (jerzygangi) [inline packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts:604]: This regex expects the page ID's thirteenth character to always be a `4` (i.e., `...[0-9a-f]{4}4[0-9a-f]{3}...`), which is an incorrect assumption. For example, `https://www.notion.so/tenant-id/fffb95d3060b80309027eb9c99605ec3` and `https://www.notion.so/fffb95d3060b80309027eb9c99605ec3` are valid Notion URL's. As such, this (and the other, similar regex's in this PR) should be fixed before approving/merging.

Additionally, I'd recommend loosening up all the Notion regex's in this PR. While validing user input strictly is a noble endeavor, these regex's make assumptions about Notion's internal infra, which is undocumented and subject to change without notice. Worse, when these assumptions prove wrong at runtime, users like [artildo](https://community.n8n.io/u/artildo) and I are left with a broken n8n Notion integration for weeks without a workaround[^1]. Holistically, the benefits of validating user input so strictly are negated by a broken Notion integration, so the regex's need to be loosened up.

[^1]: Short of making raw HTTP requests, which defeats the purpose of n8n.

File: packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts [hunk 2]
```diff
@@ -601,15 +601,15 @@ export const databasePageFields: INodeProperties[] = [
...
 							errorMessage: 'Not a valid Notion Database Page URL',
 						},
 					},
 				],
 				extractValue: {
 					type: 'regex',
 					regex:
-						'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{2,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12})',
+						'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{1,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12})',
```
Comment (jerzygangi) [inline packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts:612]: (Same comment applies here.)

File: packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts [hunk 3]
```diff
@@ -1070,15 +1070,15 @@ export const databasePageFields: INodeProperties[] = [
 						type: 'regex',
 						properties: {
 							regex:
-								'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{2,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12}).*',
+								'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{1,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12}).*',
```
Comment (jerzygangi) [inline packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts:1073]: (Same comment applies here.)

File: packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts [hunk 4]
```diff
@@ -1070,15 +1070,15 @@ export const databasePageFields: INodeProperties[] = [
...
 							errorMessage: 'Not a valid Notion Database Page URL',
 						},
 					},
 				],
 				extractValue: {
 					type: 'regex',
 					regex:
-						'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{2,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12})',
+						'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{1,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12})',
```
Comment (jerzygangi) [inline packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts:1081]: (Same comment applies here.)

File: packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts [hunk 5]
```diff
@@ -1161,15 +1161,15 @@ export const databasePageFields: INodeProperties[] = [
 						type: 'regex',
 						properties: {
 							regex:
-								'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12}).*',
-							errorMessage: 'Not a valid Notion Database URL',
+								'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{1,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12}).*',
```
Comment (jerzygangi) [inline packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts:1164]: (Same comment applies here.)

File: packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts [hunk 6]
```diff
@@ -1161,15 +1161,15 @@ export const databasePageFields: INodeProperties[] = [
...
+							errorMessage: 'Not a valid Notion Database Page URL',
 						},
 					},
 				],
 				extractValue: {
 					type: 'regex',
 					regex:
-						'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12})',
+						'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{1,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12})',
```
Comment (jerzygangi) [inline packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts:1172]: (Same comment applies here.)

File: packages/nodes-base/nodes/Notion/shared/descriptions/PageDescription.ts [hunk 7]
```diff
@@ -94,15 +94,15 @@ export const pageFields: INodeProperties[] = [
 						type: 'regex',
 						properties: {
 							regex:
-								'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{2,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12}).*',
-							errorMessage: 'Not a valid Notion Page URL',
+								'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{1,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12}).*',
```
Comment (jerzygangi) [inline packages/nodes-base/nodes/Notion/shared/descriptions/PageDescription.ts:97]: (Same comment applies here.)

File: packages/nodes-base/nodes/Notion/shared/descriptions/PageDescription.ts [hunk 8]
```diff
@@ -94,15 +94,15 @@ export const pageFields: INodeProperties[] = [
...
+							errorMessage: 'Not a valid Notion Database Page URL',
 						},
 					},
 				],
 				extractValue: {
 					type: 'regex',
 					regex:
-						'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{2,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12})',
+						'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{1,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12})',
```
Comment (jerzygangi) [inline packages/nodes-base/nodes/Notion/shared/descriptions/PageDescription.ts:105]: (Same comment applies here.)

File: packages/nodes-base/nodes/Notion/shared/descriptions/PageDescription.ts [hunk 9]
```diff
@@ -174,15 +174,15 @@ export const pageFields: INodeProperties[] = [
 						type: 'regex',
 						properties: {
 							regex:
-								'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{2,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12}).*',
-							errorMessage: 'Not a valid Notion Page URL',
+								'(?:https|http)://www.notion.so/(?:[a-z0-9-]{2,}/)?(?:[a-zA-Z0-9-]{1,}-)?([0-9a-f]{8}[0-9a-f]{4}4[0-9a-f]{3}[89ab][0-9a-f]{3}[0-9a-f]{12}).*',
```
Comment (jerzygangi) [inline packages/nodes-base/nodes/Notion/shared/descriptions/PageDescription.ts:177]: (Same comment applies here.)

**Review by jerzygangi — COMMENTED** [review]
@Joffcom Let me know if you've got any questions on this feedback, or would like an additional review after making changes. Thank you, JJ

File: packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts (hunk 1 above)
Comment (Joffcom) [reply packages/nodes-base/nodes/Notion/shared/descriptions/DatabasePageDescription.ts:604]: @jerzygangi the regex I believe was originally set up to make sure the ID also matches a UUIDv4 which is what Notion uses for a Page, You are not wrong through for a Database Page Notion just uses a UUID and they don't document the version like they do for the Page.

I will get these updated for the fields that do not specifically mention the v4 ID.

Page Properties: https://developers.notion.com/reference/page#page-object-properties
Database Page Properties: https://developers.notion.com/reference/database

I suspect we originally spotted the v4 type and ran with it.

**Comment by Joffcom** [comment]
@jerzygangi are the recent changes any better? This now follows what Notion documents so for page ids we use a UUIDv4 and for database page you can use any UUID. I have also included your URL example in the tests.

**Review by ShireenMissi — APPROVED** [review]
LGTM 🎉

**Comment by janober** [comment]
Got released with [`n8n@1.52.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.52.0)

**Comment by jerzygangi** [comment]
@Joffcom Sorry, I'm a bit late replying. This regex looks solid -- also, the new unit tests with sample uuid's is smart. Can't wait for this commit hit n8n.cloud users.

Separately, I wanted offer a 💯 emoji to you for the rapid turnaround time. I occasionally report bugs to startups, and it's amazing how many take months to reply & never fix the problem. After 1 or 2 ignored bug reports, I stop reporting. Conversely, the few engineering teams that do act quickly (like n8n) send a message to the customer that says "We value your time; we're building a quality product; please don't churn; we want your business." Keep up the hard work.

**Comment by suged** [comment]
Hello
I think the same problem happens with  Notion Update Page, Operation Archive, when the page is a Database Page

**Pull request merged**